File: CleanJava/app/src/debug/java/com/vascome/fogtail/developer_settings/LeakCanaryProxyImpl.java
package com.vascome.fogtail.developer_settings;

import android.app.Application;
import android.support.annotation.NonNull;
import android.support.annotation.Nullable;

import com.squareup.leakcanary.LeakCanary;
import com.squareup.leakcanary.RefWatcher;

/**
 * Created by vasilypopov on 11/23/17
 * Copyright (c) 2017 MVPJava. All rights reserved.
 */

public class LeakCanaryProxyImpl implements LeakCanaryProxy {

    @NonNull
    private final Application qualityMattersApp;

    @Nullable
    private RefWatcher refWatcher;

    public LeakCanaryProxyImpl(@NonNull Application qualityMattersApp) {
        this.qualityMattersApp = qualityMattersApp;
    }

    @Override
    public void init() {
        refWatcher = LeakCanary.install(qualityMattersApp);
    }

    @Override
    public void watch(@NonNull Object object) {
        if (refWatcher != null) {
            refWatcher.watch(object);
        }
    }
}

File: MVVMKotlin/app/src/debug/java/com/vascome/fogtail/utils/LeakCanaryProxyImpl.kt
package com.vascome.fogtail.utils

import android.app.Application

import com.squareup.leakcanary.LeakCanary
import com.squareup.leakcanary.RefWatcher

/**
 * Created by vasilypopov on 11/23/17
 * Copyright (c) 2017 MVPJava. All rights reserved.
 */

class LeakCanaryProxyImpl(private val qualityMattersApp: Application) : LeakCanaryProxy {

    private var refWatcher: RefWatcher? = null

    override fun init() {
        refWatcher = LeakCanary.install(qualityMattersApp)
    }

    override fun watch(anyObject: Any) {
        refWatcher?.watch(anyObject)
    }
}
